### PR TITLE
Remove `unreadActionCount` & Improve Global Indicators

### PR DESCRIPTION
**Pull request by marcaaron**
### Details

Removes `unreadActionCount` in favor of boolean unread status

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/226118
$ https://github.com/Expensify/App/issues/10768

### Tests

**Testing Global Unread Indicator**

1. Start a chat between User A and User B.
1. Ensure User A is not looking at the chat.
1. Leave a comment on the chat as User B.
1. Verify the global indicator (app icon or browser page title) has increased by 1
1. Leave another comment
1. Verify the global indicator has not changed
1. As User C add another comment to a chat that User A is not looking at
1. Verify the global indicator increases by 1
1. As User A visit the chat with User B
1. Verify the global indicator decreases by 1
1. As user A visit the chat with User C
1. Verify the global indicator decreases by 1

**Testing Unread LHN Status, “New Messages” badge, and New Line Indicator**

1. Start a chat between User A and User B.
1. Ensure User A is not looking at the chat.
1. Send a message as User B
1. Verify the LHN row shows in bold for User A to indicate unread messages
1. Visit the chat as User A and verify the LHN row no longer shows in bold
1. Verify there is a “----- New” line above the message sent by User B
1. Scroll up and verify that the “New Messages” button appears floating above the content.
1. Verify that tapping this button will scroll you to the bottom of the chat but not clear the “---- New” line marker.
1. Mark a comment as “unread”
1. Verify the “---- New” line marker changes to the place above the comment marked as unread
1. Mark as unread comments above and below the existing new marker
1. Verify the marker appears in the correct position each time
1. While the new marker is active leave a comment as User A
1. Verify the new marker and new messages badge both disappear
1. Mark another comment as unread.
1. Navigate away from the chat
1. Navigate back to the chat
1. Verify the chat in the LHN shows as read
1. Verify the New Line marker and new messages badge is still present
1. Navigate away one more time and back
1. Verify the new marker and new messages badge both disappear
1. Mark another comment as unread.
1. Navigate away from the chat
1. Navigate back to the chat
1. Minimize the app and reopen it
1. Verify the new marker and new message badge both disappear
1. Switch to a mobile view (small screen) and mark a message as unread
1. Tap the header to reveal the LHN
1. Verify the chat in the LHN shows as “unread”
1. Tap the chat
1. Verify the new marker and badge are present
1. Tap the header to reveal the LHN
1. Verify the chat in the LHN shows as “read”
1. Tap the chat
1. Verify the new marker and badge are no longer present.
1. Scroll up in the chat so that last comment is out of view
1. Leave a comment as User B
1. Verify the “new messages badge” shows for User A
1. Scroll down and verify the “new line marker” shows for User A in the correct position.
1. Scroll back up and leave another comment as User B
1. Scroll down as User A
1. Verify the new marker is in the same position as it was before

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps

- [ ] Verify that no errors appear in the JS console

### CODE REVIEW

**Comment by marcaaron** [comment]
Okaiee this is ready for reviews now :D

File: src/libs/UnreadIndicatorUpdater/index.js [hunk 1]
```diff
@@ -1,27 +1,20 @@
 import _ from 'underscore';
 import Onyx from 'react-native-onyx';
 import ONYXKEYS from '../../ONYXKEYS';
-import CONST from '../../CONST';
 import updateUnread from './updateUnread';
+import * as ReportUtils from '../ReportUtils';
 
-// Stash the unread action counts for each report
-const unreadActionCounts = {};
+// Stash the reports
```
Comment (cead22) [inline src/libs/UnreadIndicatorUpdater/index.js:7]:
```suggestion
```

File: src/libs/UnreadIndicatorUpdater/index.js [hunk 2]
```diff
@@ -1,27 +1,20 @@
...
-    // When we don't have an exact count we just let the user know there's something new
-    if (totalCount === 0 && _.some(unreadActionCounts, count => count === -1)) {
-        updateUnread(1);
-        return;
-    }
-
+    const totalCount = _.reduce(reports, (total, report) => total + (ReportUtils.isUnread(report) ? 1 : 0), 0);
     updateUnread(totalCount);
-}, 1000, {leading: false});
+}, 300, {leading: false});
```
Comment (cead22) [inline src/libs/UnreadIndicatorUpdater/index.js:17]: Curious why this changed from 1000 to 300

File: src/libs/UnreadIndicatorUpdater/index.js [hunk 3]
```diff
@@ -1,27 +1,20 @@
...
 const throttledUpdatePageTitleAndUnreadCount = _.throttle(() => {
-    const totalCount = _.reduce(unreadActionCounts, (total, reportCount) => total + Math.max(reportCount, 0), 0);
-
-    // When we don't have an exact count we just let the user know there's something new
-    if (totalCount === 0 && _.some(unreadActionCounts, count => count === -1)) {
-        updateUnread(1);
-        return;
-    }
-
+    const totalCount = _.reduce(reports, (total, report) => total + (ReportUtils.isUnread(report) ? 1 : 0), 0);
```
Comment (cead22) [inline src/libs/UnreadIndicatorUpdater/index.js:15]:
```suggestion
    const totalCount = _.filter(reports, ReportUtils.isUnread).length
```
NAB, just a suggestion

File: src/pages/home/report/ReportActionsView.js [hunk 4]
```diff
@@ -214,6 +212,9 @@ class ReportActionsView extends React.Component {
                 } else if (this.state.newMarkerSequenceNumber === 0) {
                     this.setState({newMarkerSequenceNumber: currentLastSequenceNumber});
                 }
+            } else if (this.state.newMarkerSequenceNumber === 0) {
+                // The report is not in view so we will update the newMarkerSequenceNumber
```
Comment (cead22) [inline src/pages/home/report/ReportActionsView.js:216]: Update to the last one so nothing shows as unread?

**Review by mountiny — COMMENTED** [review]
code looks clean, great job, going to test now!

**Review by cead22 — DISMISSED** [review]

File: src/pages/home/report/ReportActionsView.js (hunk 4 above)
Comment (marcaaron) [reply src/pages/home/report/ReportActionsView.js:216]: Ah so, the condition here is: 

- report isn't in view
- a new message appeared from someone besides the current user 
- no previous marker is set

So we will update the new marker with this `sequenceNumber`. This basically fixes a bug found by QA here: https://github.com/Expensify/App/issues/10768

File: src/libs/UnreadIndicatorUpdater/index.js (hunk 2 above)
Comment (marcaaron) [reply src/libs/UnreadIndicatorUpdater/index.js:17]: Oh no particular reason - this is leftover from testing and I think not needed.

**Comment by mountiny** [comment]
@marcaaron Tests well 🎉 

But a bit confused about these steps:

- Verify that tapping this button will scroll you to the bottom of the chat but not clear the “---- New” line marker.
- Mark a comment as “unread”
- Verify the “---- New” line marker appears
- Mark as unread comments above and below the existing new marker
- Verify the marker appears in the correct position each time
- While the new marker is active leave a comment as User A

So click the New Message, scorll down, dont errase the `---- New` line marker. Mark comment as unread and the marker should appear? Did you mean the marker should change to the place above the comment marked as unread. 

What is a bit unintuitive for me is that I mark old comment as unread and then I mark newer comment as unread but the New marker is moved to the newer comment. Should the marker not stay with the older comment marked as unread as it is technically the oldest comment "I did not read"?

**Review by mountiny — APPROVED** [review]

File: src/pages/home/report/ReportActionsView.js [hunk 5]
```diff
@@ -214,6 +212,10 @@ class ReportActionsView extends React.Component {
                 } else if (this.state.newMarkerSequenceNumber === 0) {
                     this.setState({newMarkerSequenceNumber: currentLastSequenceNumber});
                 }
+            } else if (this.state.newMarkerSequenceNumber === 0) {
+                // The report is not in view and we received a comment from another user while the new marker is not set
+                // so we wil set the new marker now.
```
Comment (cead22) [inline src/pages/home/report/ReportActionsView.js:217]: <3

**Review by cead22 — DISMISSED** [review]

File: src/pages/home/report/ReportActionsView.js (hunk 5 above)
Comment (mountiny) [inline src/pages/home/report/ReportActionsView.js:217]:
```suggestion
                // so we will set the new marker now.
```

**Review by mountiny — DISMISSED** [review]
NAB: Typo

**Comment by marcaaron** [comment]
>Mark comment as unread and the marker should appear? Did you mean the marker should change to the place above the comment marked as unread.

yes exactly this! I updated the steps to make it clearer, thanks!

>What is a bit unintuitive for me is that I mark old comment as unread and then I mark newer comment as unread but the New marker is moved to the newer comment. Should the marker not stay with the older comment marked as unread as it is technically the oldest comment "I did not read"?

No, in the case of manually marking a comment as unread the user is telling us they want to manually override the unread position.

**Comment by marcaaron** [comment]
Updated thanks for the reviews! 🙇‍♂️

**Review by danieldoglas — APPROVED** [review]
LGTM

**Review by thienlnam — APPROVED** [review]
This has been simplified quite a bit, thanks for these changes!

**Review by mountiny — APPROVED** [review]
> No, in the case of manually marking a comment as unread the user is telling us they want to manually override the unread position.

Coolio! Thanks

**Comment by marcaaron** [comment]
Did pass the tests.

**Comment by Julesssss** [comment]
Tested flawlessly for me on staging. Nice work.
Screenshot 2022-09-08 at 00 39 55
Screenshot 2022-09-08 at 00 36 24
Screenshot 2022-09-08 at 00 34 57
Screenshot 2022-09-08 at 00 24 48
Screenshot_20220908-004142

**Pull request merged**